wdl-grammar/src/tree.rs

```diff
@@ -690,3 +715,353 @@ impl fmt::Debug for SyntaxTree {
+    // it will be O(n). No allocations will ever be done. Thus, the
+    // ammortized cost of this is quite cheap.
+    Vec::from(results).into_boxed_slice()
+}
+
+/// An extension trait for [`SyntaxNode`]s, [`SyntaxToken`]s, and
+/// [`SyntaxElement`]s.
+pub trait SyntaxExt {
+    /// Returns whether `self` matches the provided element.
+    fn matches(&self, other: &SyntaxElement) -> bool;
+
+    /// Gets the siblings with tokens.
+    ///
+    /// **NOTE:** this needed because Rowan does not encapsulate this
+    /// functionality in a trait. Once wrapped here, most of the functions
+    /// provided by this extension trait can just be provided, which simplifies
+    /// the code. Generally speaking, this should just defer to the underlying
+    /// `siblings_with_tokens` method for each type.
+    fn siblings_with_tokens(&self, direction: Direction)
+    -> Box<dyn Iterator<Item = SyntaxElement>>;
+
+    /// Returns all of the siblings _before_ the current element.
+    ///
+    /// The siblings are returned in the order they were parsed.
+    fn preceding_siblings(&self) -> Box<[SyntaxElement]> {
+        let mut results = VecDeque::new();
+
+        self.siblings_with_tokens(Direction::Prev)
+            // NOTE: this `skip_while` is necessary because
+            // `siblings_with_tokens` returns the current node.
+            .skip_while(|e| self.matches(e))
+            .for_each(|e| results.push_front(e));
+
+        // NOTE: most of the time, this conversion will be O(1). Occassionally
+        // it will be O(n). No allocations will ever be done. Thus, the
+        // ammortized cost of this is quite cheap.
+        Vec::from(results).into_boxed_slice()
+    }
+
+    /// Returns all of the siblings _after_ the current element.
+    ///
+    /// The siblings are returned in the order they were parsed.
+    fn succeeding_siblings(&self) -> Box<[SyntaxElement]> {
+        let mut results = Vec::new();
+
+        self.siblings_with_tokens(Direction::Next)
+            // NOTE: this `skip_while` is necessary because
+            // `siblings_with_tokens` returns the current node.
+            .skip_while(|e| self.matches(e))
+            .for_each(|e| results.push(e));
+
+        // NOTE: this should always be O(1) and never require any additional
+        // allocations.
+        results.into_boxed_slice()
+    }
+
+    /// Gets all elements that are adjacent to a particular element (not
+    /// including the element itself). This means in both the forward and
+    /// reverse direction.
+    ///
+    /// The siblings are returned in the order they were parsed.
+    fn adjacent(&self) -> Box<[SyntaxElement]> {
+        let mut results = Vec::from(self.preceding_siblings());
+        results.extend(self.succeeding_siblings().iter().cloned());
+
+        // NOTE: this should always be O(1) and never require any additional
+        // allocations.
+        results.into_boxed_slice()
+    }
+}
```

Comment:
I forgot this comment on my last review. I think we can implement this simply as:

```rust
fn adjacent(&self) -> impl Iterator<Item = SyntaxElement> {
    self.preceding_siblings().chain(self.succeeding_siblings())
}
```